kafka-rest/src/test/java/io/confluent/kafkarest/response/JsonStreamTest.java

```diff
@@ -0,0 +1,289 @@
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import io.confluent.kafkarest.KafkaRestConfig;
+import io.confluent.kafkarest.TestUtils;
+import io.confluent.kafkarest.entities.v3.ProduceRequest;
+import io.confluent.kafkarest.exceptions.ProduceRequestTooLargeException;
+import io.confluent.kafkarest.response.JsonStream.SizeLimitEntityStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.StandardCharsets;
+import java.text.SimpleDateFormat;
+import java.util.Properties;
+import java.util.Random;
+import java.util.TimeZone;
+import java.util.UUID;
+import javax.ws.rs.BadRequestException;
+import org.junit.jupiter.api.Test;
+
+// CHECKSTYLE:OFF:ClassDataAbstractionCoupling
+class JsonStreamTest {
+  private static final Random RNG = new Random();
+  private static final String TEST_MESSAGE_TEMPLATE =
+      "{ " + "\"value\" : { " + "\"type\" : \"JSON\", " + "\"data\" : \"%s\" " + "}}";
+  private static final ObjectMapper OBJECT_MAPPER =
+      new ObjectMapper()
+          .registerModule(new GuavaModule())
+          .registerModule(new Jdk8Module())
+          .registerModule(new JavaTimeModule())
+          .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
+          .setDateFormat(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'"))
+          .setTimeZone(TimeZone.getTimeZone("UTC"));
+  private static final int TEST_DATA_SIZE = 1024 * 1024;
+
+  @Test
+  public void testReadOneProduceRequest_NoLimit() throws Exception {
+    Properties properties = new Properties();
+    properties.put(KafkaRestConfig.PRODUCE_REQUEST_SIZE_LIMIT_MAX_BYTES_CONFIG, "0");
+    KafkaRestConfig config = new KafkaRestConfig(properties);
+
+    File tempFile = createTempFile(UUID.randomUUID().toString());
+    // Use FileOutputStream and FileInputStream on the same file to simulate
+    // writing from clients and reading in server
+    try (FileOutputStream fos = new FileOutputStream(tempFile);
+        FileInputStream fis = new FileInputStream(tempFile);
+        JsonStream<ProduceRequest> jsonStream = createJsonStream(fis, config); ) {
+      FileChannel outputChannel = fos.getChannel();
+      // write a produce request message
+      outputChannel.write(
+          ByteBuffer.wrap(generateData(TEST_DATA_SIZE).getBytes(StandardCharsets.UTF_8)));
+      assertTrue(jsonStream.hasNext(), "Has one ProduceRequest");
+      assertTrue(
+          jsonStream.nextValue().getOriginalSize() > TEST_DATA_SIZE,
+          "Produce request is larger than data size");
+      assertFalse(jsonStream.hasNext(), "No more ProduceRequest");
+    }
+  }
+
+  @Test
+  public void testReadOneProduceRequest_WithinLimit() throws Exception {
+    final long sizeLimit = TEST_DATA_SIZE + 8 * 1024;
```

Comment:
The buffer size (internally - we can't do anything about it) of Jackson parser (that we use to parse ProduceRequest) is 8000 so +/-8KiB is theoretically the margin of error for the produce-record size counting logic.

In reality, this margin is very small comparing to the limit size we want to enforce (I intent to use the number somewhere around 27-30MiB - to accommodate the fact that BINARY message type can add overhead of 33% due to base64 encoding). 

However, I think it is useful for the users to know that they need to add a little buffer to the limit to avoid rejecting message of good size, I'll add this to the usage doc of the configuration parameter.

Reply:
added doc